### PR TITLE
[B2BP-115] - Workflow Deploy CMS

### PR DESCRIPTION
**Pull request by maurodandrea**
Creation of  scripts for the automatic deployment on AWS of CMS Strapi.

#### List of Changes
Creation of scripts for workflow GitHub and Dockerfile for docker build

#### Motivation and Context
Creation of Workflow GitHub for automatic deployment of CMS Strapi on AWS resources.

#### How Has This Been Tested?
Running scripts on locally PC pointing to a test AWS environment

#### Types of changes

- [X] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [X] My change not requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by kin0992** [comment]
@maurodandrea I'd update PR's title and description, because I don't think here Terraform is involved

File: .github/workflows/deploy_cms.yaml [hunk 1]
```diff
@@ -0,0 +1,74 @@
+name: Deploy CMS
+
+on:
+  push:
+        branches: 'main'
+        # Run only if there are at least one change matching the following paths
+        paths:
+          - 'apps/strapi-cms/*'
```
Comment (kin0992) [inline .github/workflows/deploy_cms.yaml:8]:
```suggestion
          - 'apps/strapi-cms/**'
```
`*` is stricter than `**`. Ref: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet

File: .github/workflows/deploy_cms.yaml [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@3df4ab11eba7bda6032a0b82a6bb43b11571feac # v4.0.0
+    
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@68827325e0b33c7199eb31dd4e31fbe9023e06e3 # v3.0.0
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@f95db51fddba0c2d1ec667646a06c2ce06100226 # v3.0.0
```
Comment (kin0992) [inline .github/workflows/deploy_cms.yaml:54]: Instead of making on your own these steps (and some of the following Docker related steps), you could use the [build-push-action](https://github.com/docker/build-push-action). What do you think?

File: .github/workflows/deploy_cms.yaml [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+      - name: Docker image Build
+        run: docker build -t cms-strapi:latest .
+
+      - name: Docker image Tag
+        run: docker tag cms-strapi:latest 949355639119.dkr.ecr.eu-south-1.amazonaws.com/cms-strapi:latest
+
+      - name: Login to ECR
+        uses: docker/login-action@343f7c4344506bcbf9b4de18042ae17996df046d # v3.0.0
+        with:
+            registry: 949355639119.dkr.ecr.eu-south-1.amazonaws.com
```
Comment (kin0992) [inline .github/workflows/deploy_cms.yaml:71]: Just a question here: should we treat the registry URL as a sensitive information and get it from the environment?

File: apps/strapi-cms/Dockerfile
Comment (kin0992) [inline apps/strapi-cms/Dockerfile:1]: Why do we have two different Dockerfiles?
Can't we just keep the one we are going to use for the build?

File: apps/strapi-cms/Dockerfile.prod [hunk 4]
```diff
@@ -0,0 +1,29 @@
+# Creating multi-stage build for production
+FROM node:18-alpine as build
+RUN apk update && apk add --no-cache build-base gcc autoconf automake zlib-dev libpng-dev vips-dev > /dev/null 2>&1
+ARG NODE_ENV=production
+ENV NODE_ENV=${NODE_ENV}
+
+WORKDIR /opt/
+COPY package.json package-lock.json ./
+RUN npm config set fetch-retry-maxtimeout 600000 -g && npm install --only=production
```
Comment (kin0992) [inline apps/strapi-cms/Dockerfile.prod:9]:
```suggestion
RUN npm config set fetch-retry-maxtimeout 600000 -g && npm install
```
if you set the `NODE_ENV` to production, this flag is not necessary

File: apps/strapi-cms/Dockerfile.prod [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+ENV PATH /opt/node_modules/.bin:$PATH
+WORKDIR /opt/app
+COPY . .
+RUN npm run build
+
+# Creating final production image
+FROM node:16-alpine
```
Comment (kin0992) [inline apps/strapi-cms/Dockerfile.prod:16]: You should use the version specified in the `.node-version` file

File: apps/strapi-cms/Dockerfile.prod [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+WORKDIR /opt/
+COPY --from=build /opt/node_modules ./node_modules
+WORKDIR /opt/app
+COPY --from=build /opt/app ./
+ENV PATH /opt/node_modules/.bin:$PATH
+
+RUN chown -R node:node /opt/app
+USER node
+EXPOSE 1337
+CMD ["npm", "run", "start"]
```
Comment (kin0992) [inline apps/strapi-cms/Dockerfile.prod:29]: You shouldn't use `npm` script within the `CMD` instruction.

Check this [article with some best practices about containerize nodejs applications with docker](https://snyk.io/blog/10-best-practices-to-containerize-nodejs-web-applications-with-docker/) 😃

**Review by kin0992 — CHANGES_REQUESTED** [review]

File: .github/workflows/deploy_cms.yaml (hunk 3 above)
Comment (maurodandrea) [reply .github/workflows/deploy_cms.yaml:71]: Yes, we can use ${{ steps.login-ecr.outputs.registry }} for reference to url registry

File: .github/workflows/deploy_cms.yaml (hunk 2 above)
Comment (maurodandrea) [reply .github/workflows/deploy_cms.yaml:54]: yes, I had seen this suggested action. It does not save us these steps (docker/setup-qemu-action, setup-buildx-action, Login to ECR) but it includes the build, tag and push steps in a single action. 
Example:
  - name: Build Push and Tag
    uses: docker/build-push-action@v4
    with:
      context: .
      file: ./Dockerfile
      push: true
      tags: ${{ steps.login-ecr.outputs.registry }}/cms-strapi:latest
      provenance: false

what do you think about it ?

File: apps/strapi-cms/Dockerfile
Comment (maurodandrea) [reply apps/strapi-cms/Dockerfile:1]: Yes, fair observation. I didn't know which mode Strapi should be deployed, so I put 2 dockerfiles. Now I have had confirmation that it will be deployed in production mode, so there will only be one Dockerfile per production

File: .github/workflows/deploy_cms.yaml (hunk 2 above)
Comment (kin0992) [reply .github/workflows/deploy_cms.yaml:54]: I found [an old project we worked on](https://github.com/pagopa/pn-local-emulator/blob/1771071b3274203eec83636c55733984311158c9/.github/workflows/main.yaml#L54-L57), where we have a similar use case: we need to create and push a Docker image and there we used only the build-push-action I mentioned.

I'm sure that something could be evolved, indeed we used `v3` of that action while now the major version is `v5`. But, could you investigate a little to see if there is possible to omit those steps that are using setup-qemu-action and setup-buildx-action?

File: apps/strapi-cms/Dockerfile.prod (hunk 6 above)
Comment (maurodandrea) [reply apps/strapi-cms/Dockerfile.prod:29]: "CMD npm run start", right ?

File: apps/strapi-cms/Dockerfile
Comment (kin0992) [reply apps/strapi-cms/Dockerfile:1]: In this case I'd keep only the file called `Dockerfile`

Comment (maurodandrea) [reply apps/strapi-cms/Dockerfile:1]: Yes

File: apps/strapi-cms/Dockerfile.prod (hunk 6 above)
Comment (maurodandrea) [reply apps/strapi-cms/Dockerfile.prod]: change made

File: apps/strapi-cms/Dockerfile
Comment (maurodandrea) [reply apps/strapi-cms/Dockerfile:1]: change made

File: .github/workflows/deploy_cms.yaml (hunk 2 above)
Comment (maurodandrea) [reply .github/workflows/deploy_cms.yaml]: change present in the latest commit

File: apps/strapi-cms/Dockerfile.prod (hunk 6 above)
Comment (maurodandrea) [reply apps/strapi-cms/Dockerfile.prod]: change present in the latest commit

File: .github/workflows/deploy_cms.yaml [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@3df4ab11eba7bda6032a0b82a6bb43b11571feac # v4.0.0
+    
+      #- name: Set up QEMU
+      #  uses: docker/setup-qemu-action@68827325e0b33c7199eb31dd4e31fbe9023e06e3 # v3.0.0
+
+      #- name: Set up Docker Buildx
+      #  uses: docker/setup-buildx-action@f95db51fddba0c2d1ec667646a06c2ce06100226 # v3.0.0
```
Comment (kin0992) [inline .github/workflows/deploy_cms.yaml:54]:
```suggestion
```

File: .github/workflows/deploy_cms.yaml [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+      #- name: Docker image Tag
+      #  run: docker tag cms-strapi:latest 949355639119.dkr.ecr.eu-south-1.amazonaws.com/cms-strapi:latest
+
+      #- name: Login to ECR
+      #  uses: docker/login-action@343f7c4344506bcbf9b4de18042ae17996df046d # v3.0.0
+      #  with:
+      #      registry: 949355639119.dkr.ecr.eu-south-1.amazonaws.com
+    
+      #- name: Docker image Push
+      #  run: aws ecr get-login-password --region eu-south-1 | docker push 949355639119.dkr.ecr.eu-south-1.amazonaws.com/cms-strapi:latest
```
Comment (kin0992) [inline .github/workflows/deploy_cms.yaml:74]:
```suggestion
```

File: .github/workflows/deploy_cms.yaml [hunk 9]
```diff
@@ -0,0 +1,90 @@
+name: Deploy CMS
+
+on:
+  push:
+        branches: 'main'
```
Comment (kin0992) [inline .github/workflows/deploy_cms.yaml:5]:
```suggestion
        branches: ['main']
```
`branches` need an array, so you can do as I suggested or even
```
branches:
    - branch1
    - branch2
```

File: .github/workflows/deploy_cms.yaml [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    inputs:
+      logLevel:
+        description: 'Log level'
+        required: true
+        default: 'warning'
+        type: choice
+        options:
+        - info
+        - warning
+        - debug
```
Comment (kin0992) [inline .github/workflows/deploy_cms.yaml:21]:
```suggestion
        options:
        - warning
```
or maybe we can delete `logLevel` from this workflow

File: .github/workflows/deploy_cms.yaml [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    shell: bash
+    working-directory: apps/strapi-cms
+
+permissions:
+  id-token: write   # This is required for requesting the JWT
+  contents: read    # This is required for actions/checkout
+
+jobs:
+  deploy:
+    name: Docker CMS
```
Comment (kin0992) [inline .github/workflows/deploy_cms.yaml:41]:
```suggestion
    name: Build and push CMS image
```
what do you think?

File: apps/strapi-cms/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,29 @@
+# Creating multi-stage build for production
+FROM node:18-alpine as build
```
Comment (kin0992) [inline apps/strapi-cms/Dockerfile:2]: you should get a version using a SHA.
Moreover, I think you should set the version to a variable and use the very same version (avoiding duplication) also in the second layer of this multi-stage build.

I would use the version defined in the [`.node-version` file](https://github.com/pagopa/b2b-portals/blob/main/.node-version)

Comment (maurodandrea) [reply apps/strapi-cms/Dockerfile]: ok, with last commit i have insert var and SHA by new row:  "ARG NODE_IMAGE=node:18-alpine@sha256:d51f2f5ce2dc7dfcc27fc2aa27a6edc66f6b89825ed4c7249ed0a7298c20a45a"
The node version defined in the [.node-version file](https://github.com/pagopa/b2b-portals/blob/main/.node-version) is not presente in DockerHub image: https://hub.docker.com/_/node/tags?page=1&name=18.7.1
The image "node:18.7-alpine" has multiple vulnerabilities, strapi recommends using the "node:18-alpine" image.
Furthermore, inserting the SHA we cannot bind the docker file to the node version defined in the [.node-version file](https://github.com/pagopa/b2b-portals/blob/main/.node-version)

File: apps/strapi-cms/Dockerfile
Comment (kin0992) [inline apps/strapi-cms/Dockerfile:1]: Have you tried locally to build the image?

I'm using podman (but I don't think this is a problem), and I'm having this issue:
```
> strapi-cms@0.0.0 build
> strapi build

tsconfig.json:2:14 - error TS6053: File '../../tsconfig' not found.

2   "extends": "../../tsconfig",
               ~~~~~~~~~~~~~~~~

Found 1 error(s).

Error: building at STEP "RUN npm run build": while running runtime: exit status 1
```

For your information, I used these commands:
- `podman build -t b2b-portals .` executed in `/apps/strapi-cms`
- `podman build -t b2b-portals ./apps/strapi-cms/` executed in the root of the repository
both are giving me the same error

File: .github/workflows/deploy_cms.yaml [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        uses: docker/build-push-action@0565240e2d4ab88bba5387d719585280857ece09 # v5.0.0
+        env:
+          ECR_REGISTRY: ${{ steps.login-ecr.outputs.registry }}
+          ECR_REPOSITORY: strapi
+          IMAGE_TAG: latest
+        with:
+            context: apps/strapi-cms
+            push: true
+            tags: ${{ steps.login-ecr.outputs.registry }}/strapi:latest
+            provenance: false
```
Comment (kin0992) [inline .github/workflows/deploy_cms.yaml:70]: What do you think to push images using tags?
Check [this workflow](https://github.com/pagopa/juggler/blob/7572f43e1f472877934ddfbaac6063bf24d95109/.github/workflows/push-image.yaml#L37-L47).

The idea is to push an image with a SHA or a tag. Note that this require a change in the workflow trigger.

File: .github/workflows/deploy_cms.yaml [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+    steps:
+      - name: Checkout
+        uses: actions/checkout@3df4ab11eba7bda6032a0b82a6bb43b11571feac # v4.0.0
+    
+
+      - name: Configure AWS Credentials
+        uses: ./.github/actions/configure-aws-credentials
+        with:
+          aws_region: ${{ env.AWS_REGION || 'eu-south-1' }}
+          role_to_assume: ${{ secrets.IAM_ROLE }}
```
Comment (kin0992) [inline .github/workflows/deploy_cms.yaml:53]: we could create a new IAM_ROLE to follow [the least privilege principle](https://en.wikipedia.org/wiki/Principle_of_least_privilege)

**Review by kin0992 — CHANGES_REQUESTED** [review]

File: apps/strapi-cms/Dockerfile
Comment (maurodandrea) [reply apps/strapi-cms/Dockerfile:1]: Yes I tested it locally and it gives the same error. However, if you delete the "extends" line from the tsconfig and everything works. Precisely for this reason I am talking with Simone to try to understand how to solve it. Do you have any suggestions on this? Thank you @kin0992 @datalek

File: .github/workflows/deploy_cms.yaml (hunk 14 above)
Comment (maurodandrea) [reply .github/workflows/deploy_cms.yaml]: ok, i have open a PR https://github.com/pagopa/b2b-portals/pull/58 for create new role: 
1) GitHubActionDeployECS for deploy CMS on ECS. In Workflow i have write secrets.IAM_ROLE_DEPLOY_CMS
2) GitHubActionDeployWebsite for deploy a static website on S3. In Workflow i have write secrets.IAM_ROLE_DEPLOY_WEBSITE

File: .github/workflows/deploy_cms.yaml (hunk 13 above)
Comment (maurodandrea) [reply .github/workflows/deploy_cms.yaml]: ok, i've set the IMAGE_TAG: ${{ github.sha }}

File: .infrastructure/99_outputs.tf
Comment (kin0992) [inline .infrastructure/99_outputs.tf:1]: why this change is in this PR?

Comment (maurodandrea) [reply .infrastructure/99_outputs.tf:1]: has been added to show the loadbalancer URL in output to connect to the CMS console. If you want, we can also remove this part of code and obtain the URL from the AWS console. Since it is related to the CMS, I included it in this PR

File: .github/workflows/deploy_cms.yaml [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+        with:
+          aws_region: ${{ env.AWS_REGION || 'eu-south-1' }}
+          role_to_assume: ${{ secrets.IAM_ROLE_DEPLOY_CMS }}
+
+      - name: Login to Amazon ECR
+        id: login-ecr
+        uses: aws-actions/amazon-ecr-login@062b18b96a7aff071d4dc91bc00c4c1a7945b076 # v2.0.1
+
+      - name: Build Push and Tag
+        uses: docker/build-push-action@0565240e2d4ab88bba5387d719585280857ece09 # v5.0.0
```
Comment (kin0992) [inline .github/workflows/deploy_cms.yaml:59]:
```suggestion
        uses: docker/build-push-action@4a13e500e55cf31b7a5d59a38ab2040ab0f42f56 # v5.1.0
```
I think this will work too, but we need to test it. Anyway, I'd use latest version

File: .github/workflows/deploy_cms.yaml [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+        env:
+          ECR_REGISTRY: ${{ steps.login-ecr.outputs.registry }}
+          ECR_REPOSITORY: strapi
+          IMAGE_TAG: ${{ github.sha }}
+        with:
+            context: apps/strapi-cms
+            push: true
+            tags: ${{ steps.login-ecr.outputs.registry }}/strapi:${{ github.sha }}
```
Comment (kin0992) [inline .github/workflows/deploy_cms.yaml:67]: should we get the hardcoded value of `strapi` from the envinronment?

File: .github/workflows/deploy_cms.yaml (hunk 15 above)
Comment (maurodandrea) [reply .github/workflows/deploy_cms.yaml:59]: yes, the release v5.1.0 was released last month, so I consider it reliable

File: .github/workflows/deploy_cms.yaml (hunk 16 above)
Comment (maurodandrea) [reply .github/workflows/deploy_cms.yaml:67]: right, with next commit set tags: ${{ env.ECR_REGISTRY }}/${{ env.ECR_REPOSITORY }}:${{ env.IMAGE_TAG }}

**Comment by maurodandrea** [comment]
@datalek: Can you perform the merge because this operation is disabled for me? Thanks

**Comment by datalek** [comment]
> @datalek: Can you perform the merge because this operation is disabled for me? Thanks

The merge is not available because there are some suggestions from @kin0992  that haven't been addressed
